### PR TITLE
Terraform Template for the AWS Codepipeline Implementation

### PR DESCRIPTION
**Pull request by jasonquinto**
Hi @rafaljanicki this is the Terraform template for my AWS CodePipeline implementation. You might notice that it’s a single template, and I didn't create separate templates for the build and upload pipelines. The reason is that I thought it could be simplified by reusing this template and specifying in the `.tfvars` which pipeline it will be used for, using appropriate naming schemes in CodeBuild and Codepipeline (e.g., `build` or `upload`).

### CODE REVIEW

File: cloudwatch.tf [hunk 1]
```diff
@@ -0,0 +1,13 @@
+resource "aws_cloudwatch_log_group" "codebuild_log_group" {
+  name              = "/aws/codebuild/${local.project_name_codebuild}"
+  retention_in_days = 90  # Log retention set to 90 days
```
Comment (rafaljanicki) [inline cloudwatch.tf]: Make the log retention a variable (also no need for a comment that's basically the argument name + value ;) )

File: codebuild.tf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  environment {
+    compute_type                = "BUILD_GENERAL1_MEDIUM"  # Compute resources
+    image                       = "aws/codebuild/amazonlinux2-x86_64-standard:5.0"  # Build environment image
+    type                        = "LINUX_CONTAINER"  # Environment type
+    image_pull_credentials_type = "CODEBUILD"
+    privileged_mode             = false  # Disable privileged mode
+
+    environment_variable {
+      name  = "ENV"  # Environment variable
```
Comment (rafaljanicki) [inline codebuild.tf]: If it's not needed, don't add the `environment_variable` block

File: codepipeline.tf [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    action {
+      name             = "SourceAction"  # Source stage action name
+      category         = "Source"
+      owner            = "AWS"
+      provider         = "CodeStarSourceConnection"  # GitHub connection
+      version          = "1"
+      configuration = {
+        ConnectionArn   = var.connection_arn  # GitHub connection ARN
+        FullRepositoryId = var.repository_name  # Full GitHub repository ID
```
Comment (rafaljanicki) [inline codepipeline.tf]: Formatting

File: iam.tf [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        ],
+        Resource = "*"
+      },
+      {
+        Effect   = "Allow",
+        Action   = [
+          "s3:GetObject",
+          "s3:PutObject"
+        ],
+        Resource = "*"
```
Comment (rafaljanicki) [inline iam.tf]: Limit to specific buckets

File: iam.tf [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      },
+      {
+        Effect   = "Allow",
+        Action   = [
+          "ssm:GetParameter",
+          "ssm:GetParameters",
+          "ssm:GetParametersByPath"
+        ],
+        Resource = "*"
```
Comment (rafaljanicki) [inline iam.tf]: Limit that as well

File: iam.tf [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      },
+      {
+        Effect   = "Allow",
+        Action   = [
+          "s3:GetObject",
+          "s3:PutObject"
+        ],
+        Resource = "*"
```
Comment (rafaljanicki) [inline iam.tf]: Limit

File: iam.tf [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      },
+      {
+        Effect   = "Allow",
+        Action   = [
+          "codepipeline:*"
```
Comment (rafaljanicki) [inline iam.tf]: Limit?

File: iam.tf [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    Statement = [
+      {
+        Effect = "Allow",
+        Action = [
+          "kms:Decrypt",
+          "kms:Encrypt",
+          "kms:GenerateDataKey",
+          "kms:DescribeKey"
+        ],
+        Resource = "*"  # Replace with specific KMS key ARN if needed
```
Comment (rafaljanicki) [inline iam.tf]: Do replace

File: s3.tf [hunk 9]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (rafaljanicki) [inline s3.tf]: Formatting

File: s3.tf [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  tags = var.tags  # Apply tags
+}
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "codebuild_artifacts_encryption" {
+  bucket = aws_s3_bucket.codebuild_artifacts.bucket
+
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm     = "aws:kms"  # Use AWS KMS for encryption
+      kms_master_key_id = "alias/aws/s3"  # Default KMS key for S3
```
Comment (rafaljanicki) [inline s3.tf]: No need to mention it if you're using the default one

File: s3.tf [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+resource "aws_s3_bucket_lifecycle_configuration" "codebuild_artifacts_lifecycle" {
+  bucket = aws_s3_bucket.codebuild_artifacts.bucket
+
+  rule {
+    id     = "lifecycle-rule"  # Lifecycle rule ID
+    status = "Enabled"  # Enable the lifecycle rule
+
+    expiration {
+      days = 90  # Expire objects after 90 days
```
Comment (rafaljanicki) [inline s3.tf]: I don't think that's a good idea - what if I'd like to rerun the pipeline after 3 months?

File: s3.tf [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+      noncurrent_days = 30  # Expire noncurrent versions after 30 days
+    }
+  }
+}
+
+resource "aws_s3_bucket_versioning" "codebuild_artifacts_versioning" {
+  bucket = aws_s3_bucket.codebuild_artifacts.id
+
+  versioning_configuration {
+    status = "Enabled"  # Enable versioning
```
Comment (rafaljanicki) [inline s3.tf]: I don't think it's needed - every artifact has a unique name, right?

File: terraform.tfvars [hunk 13]
```diff
@@ -0,0 +1,16 @@
+region              = "$ACTUAL_VALUE"  # AWS region for deployment
```
Comment (rafaljanicki) [inline terraform.tfvars]: Just don't include that file in the repo instead of a stub file like that :)

**Comment by rafaljanicki** [comment]
This looks quite great! I've added a number of comments, but these are mostly around smaller things. Good job!

Next time we'll learn Terragrunt to make it easier to provide values

**Review by rafaljanicki — CHANGES_REQUESTED** [review]

File: iam.tf [hunk 14]
```diff
@@ -31,22 +32,25 @@ resource "aws_iam_role_policy" "codebuild_policy" {
...
       },
       {
         Effect   = "Allow",
         Action   = [
           "s3:GetObject",
           "s3:PutObject"
         ],
-        Resource = "*"
+        Resource = [
+          "${aws_s3_bucket.codebuild_artifacts.arn}",
```
Comment (rafaljanicki) [inline iam.tf]: For Get & Put objects, you can just leave the `"${aws_s3_bucket.codebuild_artifacts.arn}/*"` line

File: variable.tf [hunk 15]
```diff
@@ -48,6 +33,16 @@ variable "tags" {
   type        = map(string)
 }
 
+variable "log_retention_days" {
+  description = "Retention period for CloudWatch logs (in days)"
+  type        = number
+}
+
+variable "parameter_name" {
+  description = "The name of the SSM Parameter that needs to be retrieved"
```
Comment (rafaljanicki) [inline variable.tf]: What this parameter is for? Let's make that a bit more descriptive

**Review by rafaljanicki — CHANGES_REQUESTED** [review]
Last two bits

File: iam.tf [hunk 16]
```diff
@@ -55,22 +59,32 @@ resource "aws_iam_role_policy" "codebuild_policy" {
...
+        Resource = "arn:aws:ssm:${var.region}:${data.aws_caller_identity.current.account_id}:parameter/${var.parameter_name}"
       },
       {
         Effect   = "Allow",
         Action   = [
           "ecr:BatchGetImage",
           "ecr:GetDownloadUrlForLayer"
         ],
-        Resource = "*"
+        Resource = "arn:aws:ecr:${var.region}:${data.aws_caller_identity.current.account_id}:repository/${var.environment}-${var.pipeline_name}-codebuild*"
```
Comment (rafaljanicki) [inline iam.tf]: Where is that repository created?

**Review by rafaljanicki — CHANGES_REQUESTED** [review]
Last two bits

File: iam.tf (hunk 16 above)
Comment (jasonquinto) [reply iam.tf]: It depends. The template doesn't create the repository, it assumes that you already have an ECR repository and only requires you to provide its URI. Should I add the creation of the ECR repository to the template?

File: variable.tf (hunk 15 above)
Comment (jasonquinto) [reply variable.tf]: It's for the variables used by the pipeline. Sorry for the vague description, I will revise this.

File: iam.tf (hunk 14 above)
Comment (jasonquinto) [reply iam.tf]: Got it. I will include it in my next revision.

File: iam.tf (hunk 16 above)
Comment (rafaljanicki) [reply iam.tf]: I believe so, yes

**Comment by jasonquinto** [comment]
@rafaljanicki Hey, I've added the creation of ECR resource and policy plus the other changes requested.

**Review by rafaljanicki — APPROVED** [review]

**Pull request merged**